[For discussion][DeviceMesh] Use a shared_state to cache pg per layout, root_mesh and rank_map

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #166010
* #167753
* #167590
* #167581

To avoid creating extra PGs when not needed, (for example when calling unflatten many times and some dims share the same layout), we want to create a pg cache mechanism. We cache a PG by a pair of layout (_MeshLayout) and its pg_option, so if users flatten or unflatten into the same layout+pg option (by default pg_option will be None if not set, that way the cache key falls back to layout itself) we will not create process group if already created.

Also to further consolidate the bookkeeping of DM, we created a device_type, shared_state to store root_mesh, rank_map and pg caches. This way all shared info for a given device mesh universe now becomes a singleton.

cc @H-Huang @awgu @wanchaol @fegin @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk

fduwjj committed:
[WIP][DeviceMesh] Use a shared_state to cache pg per layout, root_mesh and rank_map

[ghstack-poisoned]

fduwjj committed:
Update on "[WIP][DeviceMesh] Use a shared_state to cache pg per layout, root_mesh and rank_map"




cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[For discussion][DeviceMesh] Use a shared_state to cache pg per layout, root_mesh and rank_map"


We want to create a shared_state to store root_mesh, rank_map and pg caches. We can add more into it down the road, so that it becomes a singleton for bookkeeping and also align with our original proposal to move toward the idea of mesh universe.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[For discussion][DeviceMesh] Use a shared_state to cache pg per layout, root_mesh and rank_map"


We want to create a shared_state to store root_mesh, rank_map and pg caches. We can add more into it down the road, so that it becomes a singleton for bookkeeping and also align with our original proposal to move toward the idea of mesh universe.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -123,6 +123,44 @@ def _get_device_handle(device_type: str = "cuda"):
         """
         return getattr(torch, device_type, None)
 
+    class _SharedState:
```
This is the designed proposed by Luca in https://github.com/pytorch/pytorch/pull/163213#discussion_r2387806468 and I think it makes sense. Now this PR is trying to do both: pg caching and shared state. And each universe will share the same state.

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -463,11 +520,15 @@ def _init_process_groups(
 
         def _get_root_mesh(self) -> "DeviceMesh":
-            return self._root_mesh if self._root_mesh else self
+            return not_none(self._shared_state.get_root_mesh())
```
This is the only instance where we're accessing the root mesh of the shared state. In principle, once we introduce the concept of shared state, we can get rid of the concept of root meshes.

What is preventing us from simply removing this `_get_root_mesh` method?

fduwjj replied:
We might need more PRs to remove `_get_root_mesh` not in this PR.

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -125,6 +125,44 @@ def _get_device_handle(device_type: str = "cuda"):
+            self._rank_map = rank_map
+            self._root_mesh = root_mesh
+            self._backend_cache: dict[_MeshLayout, str] = {}
+            self._lock = threading.Lock()
```
Just like @wconstab, I'm also very puzzled by the fact that we need a lock. We've never needed that before, even though we already knew that DeviceMeshes were being accessed by multiple threads (notably: the main thread during forward and the autograd thread pool during backward). And yet, we have never encountered race conditions that motivated introducing locks.

What has changed?

fduwjj replied:
I mean we can remove it and see if it works. I don't have a strong opinion on this.

wconstab replied:
well, in this case i'm pretty sure we don't need them.  but in general, its not a good approach to add / remove locks and see if it works- you have to have a design and justify whether you need locks, it should not be guesswork!

fduwjj committed:
Update on "[For discussion][DeviceMesh] Use a shared_state to cache pg per layout, root_mesh and rank_map"


We want to create a shared_state to store root_mesh, rank_map and pg caches. We can add more into it down the road, so that it becomes a singleton for bookkeeping and also align with our original proposal to move toward the idea of mesh universe.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

lw reviewed (commented):
Thanks!

I think this PR is a great starting point for the discussion around PG caching, and defining a clear mental model for how different DeviceMesh should share state.

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -238,6 +277,11 @@ def __init__(
+            if _shared_state is None:
+                self._shared_state = _SharedState(rank_map, self)
+            else:
+                self._shared_state = _shared_state
```
yes it was merged sorry I just updated it yesterday.

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -367,26 +411,39 @@ def _init_process_groups(
-                        backend="cpu:gloo,cuda:nccl",
-                        ranks=ranks,
-                        group_desc="mesh_default",
+                backend_cache = shared_state.get_backend_from_cache(layout)
```
sure, let me try to make a preliminary PR.

fduwjj committed:
Update on "[For discussion][DeviceMesh] Use a shared_state to cache pg per layout, root_mesh and rank_map"


We want to create a shared_state to store root_mesh, rank_map and pg caches. We can add more into it down the road, so that it becomes a singleton for bookkeeping and also align with our original proposal to move toward the idea of mesh universe.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[For discussion][DeviceMesh] Use a shared_state to cache pg per layout, root_mesh and rank_map"


We want to create a shared_state to store root_mesh, rank_map and pg caches. We can add more into it down the road, so that it becomes a singleton for bookkeeping and also align with our original proposal to move toward the idea of mesh universe.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[For discussion][DeviceMesh] Use a shared_state to cache pg per layout, root_mesh and rank_map"


We want to create a shared_state to store root_mesh, rank_map and pg caches. We can add more into it down the road, so that it becomes a singleton for bookkeeping and also align with our original proposal to move toward the idea of mesh universe.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[For discussion][DeviceMesh] Use a shared_state to cache pg per layout, root_mesh and rank_map"


We want to create a shared_state to store root_mesh, rank_map and pg caches. We can add more into it down the road, so that it becomes a singleton for bookkeeping and also align with our original proposal to move toward the idea of mesh universe.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

fduwjj committed:
Update on "[For discussion][DeviceMesh] Use a shared_state to cache pg per layout, root_mesh and rank_map"


We want to create a shared_state to store root_mesh, rank_map and pg caches. We can add more into it down the road, so that it becomes a singleton for bookkeeping and also align with our original proposal to move toward the idea of mesh universe.

cc H-Huang awgu wanchaol fegin wz337 wconstab d4l3k pragupta msaroufim dcci

[ghstack-poisoned]

kwen2501 commented on test/distributed/test_device_mesh.py:
```diff
@@ -1051,6 +1051,26 @@ def test_unflatten_mesh_3d(self):
+            mesh_dim_names=("world",),
+        )
+        non_ep_mesh = global_mesh._unflatten(0, (2, 2, 2), ("dp", "cp", "tp"))
+        ep_mesh = global_mesh._unflatten(0, (2, 2, 2), ("dp", "ep", "ep_tp"))
```
What's the difference between these two lines vs a user giving multiple names to a dimension? Or a complex name such as "cp_or_ep"?
